src/charm.py

```diff
@@ -1,217 +1,416 @@
-     service_key: '{pagerduty_key}'
-"""
-
-HA_PORT = "9094"
-UNIT_ADDRESS = "{}-{}.{}-endpoints.{}.svc.cluster.local"
-
-
-class BlockedStatusError(Exception):
-    pass
-
-
-def status_catcher(func):
-    @functools.wraps(func)
-    def new_func(self, *args, **kwargs):
+from ops.framework import StoredState
+from ops.model import ActiveStatus, MaintenanceStatus
+
+import urllib.parse
+import requests
+import yaml
+import json
+import logging
+from typing import Dict, List, Optional, Any
+
+logger = logging.getLogger(__name__)
+
+
+class AlertmanagerAPIClient:
+    """Alertmanager HTTP API client.
+    """
+    def __init__(self, address: str, port: int):
+        self.base_url = "http://{}:{}/".format(address, port)
+
+    def reload(self) -> bool:
+        """Send a POST request to to hot-reload the config.
+        This reduces down-time compared to restarting the service.
+
+        Returns:
+          True if reload succeeded (returned 200 OK); False otherwise.
+        """
+        url = urllib.parse.urljoin(self.base_url, "/-/reload")
         try:
-            func(self, *args, **kwargs)
-        except BlockedStatusError as e:
-            self.unit.status = BlockedStatus(str(e))
+            response = requests.post(url, timeout=2.0)
+            logger.debug("config reload via %s: %d %s",
+                         url, response.status_code, response.reason)
+            return response.status_code == 200 and response.reason == 'OK'
+        except (requests.exceptions.ConnectionError, requests.exceptions.ConnectTimeout) as e:
+            logger.debug("config reload error via %s: %s", url, str(e))
+            return False
+
+    def status(self) -> Optional[dict]:
+        url = urllib.parse.urljoin(self.base_url, "/api/v2/status")
+        try:
+            response = requests.get(url)
+            if response.ok and response.status_code == 200:
+                status = json.loads(response.text)
+            else:
+                status = None
+        except (requests.exceptions.ConnectionError, requests.exceptions.ConnectTimeout):
+            status = None
 
-    return new_func
+        return status
 
 
 class AlertmanagerCharm(CharmBase):
+    _container_name: str = "alertmanager"  # automatically determined from charm name
+    _layer_name = "alertmanager"  # layer label argument for container.add_layer
+    _service_name: str = "alertmanager"  # chosen arbitrarily to match charm name
+    _peer_relation_name: str = "replicas"  # must match metadata.yaml peer role name
+    _api_port: int = 9093  # port to listen on for the web interface and API
+    _ha_port: int = 9094  # port for HA-communication between multiple instances of alertmanager
+
+    # path, inside the workload container, to the alertmanager configuration file
+    _config_path = "/etc/alertmanager/alertmanager.yml"
+
+    # path, inside the workload container for alertmanager logs, e.g. 'nflogs', 'silences'.
+    _storage_path = "/alertmanager"
```

Comment:
I think the more common python pattern would be to make these globals. This works too though

Reply:
Yeah I had it as globals at first, but module-wide variables for this seem like an anti-pattern to me tbh.
1. I needed them in the provider and tests, so `self.charm._api_port` and `self.harness.charm._api_port` seem more correct than `from charm import API_PORT`.
2. Technically, a single `charm.py` could have two different charm classes. Not that we'll ever do it.

Reply:
The library should not be calling in to charm code. Even if it is just a constant. I made another comment about this.

Reply:
Is `from charm import API_PORT` not "charm code"?